### ISSUE TITLE
Deterministic behavior of Faster RCNN

### ISSUE THREAD

**Issue by xXBasti** [pytorch/vision#1058]
Hi I am trying to run the FasterRCNN in a deterministic manner.
I am using a fixed seed and also set deterministic to True, basiclly followed the instructions on pytorch.org and the forum.

def setup_random(seed=42):
    np.random.seed(seed)
    random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

    #torch.backends.cudnn.enabled = False
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

def init_fn_worker(seed=42):
    np.random.seed(seed)
    random.seed(seed)

My model is basicly created by 
model = FasterRCNN.fasterrcnn_resnet50_fpn(pretrained=False, num_classes=num_classes)

and I tested this on the Kitti 2D Object detection dataset.

CUDA 10.1
CUDNN 7
Nvidia 418.40.4

I installed pytorch 1.1 with cuda 10 with pip3 (pytorch website).
For this test I also deactivated warmup_lr_scheduler from the engine file and set num_workers to 0.
After training 1 Epoch, I get the following results:

First run:

Epoch: [0]  [  0/400]  eta: 0:07:45  lr: 0.001000  loss: 2.1258 (2.1258)  loss_classifier: 1.6756 (1.6756)  loss_box_reg: 0.3734 (0.3734)  loss_objectness: 0.0559 (0.0559)  loss_rpn_box_reg: 0.0209 (0.0209)  time: 1.1646  data: 0.3258  max mem: 4116
Epoch: [0]  [100/400]  eta: 0:03:03  lr: 0.001000  loss: 0.8152 (1.1617)  loss_classifier: 0.3949 (0.6956)  loss_box_reg: 0.3512 (0.3663)  loss_objectness: 0.0159 (0.0624)  loss_rpn_box_reg: 0.0189 (0.0374)  time: 0.6120  data: 0.0122  max mem: 5492
Epoch: [0]  [200/400]  eta: 0:02:04  lr: 0.001000  loss: 0.7226 (0.9637)  loss_classifier: 0.3138 (0.5190)  loss_box_reg: 0.3706 (0.3659)  loss_objectness: 0.0294 (0.0467)  loss_rpn_box_reg: 0.0227 (0.0321)  time: 0.6444  data: 0.0123  max mem: 5495
Epoch: [0]  [300/400]  eta: 0:01:02  lr: 0.001000  loss: 0.6809 (0.8708)  loss_classifier: 0.3010 (0.4415)  loss_box_reg: 0.3293 (0.3507)  loss_objectness: 0.0179 (0.0455)  loss_rpn_box_reg: 0.0234 (0.0331)  time: 0.6351  data: 0.0129  max mem: 5504
Epoch: [0]  [399/400]  eta: 0:00:00  lr: 0.001000  loss: 0.6667 (0.8172)  loss_classifier: 0.2803 (0.4020)  loss_box_reg: 0.2960 (0.3403)  loss_objectness: 0.0214 (0.0424)  loss_rpn_box_reg: 0.0219 (0.0325)  time: 0.6133  data: 0.0112  max mem: 5504
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.158
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.334
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.120
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.097
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.166
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.196
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.149
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.332
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.358
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.309
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.351
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.411

Second run:

Epoch: [0]  [  0/400]  eta: 0:07:51  lr: 0.001000  loss: 2.1258 (2.1258)  loss_classifier: 1.6756 (1.6756)  loss_box_reg: 0.3734 (0.3734)  loss_objectness: 0.0559 (0.0559)  loss_rpn_box_reg: 0.0209 (0.0209)  time: 1.1798  data: 0.3474  max mem: 4116
Epoch: [0]  [100/400]  eta: 0:03:11  lr: 0.001000  loss: 0.8089 (1.1623)  loss_classifier: 0.4004 (0.6956)  loss_box_reg: 0.3541 (0.3669)  loss_objectness: 0.0159 (0.0624)  loss_rpn_box_reg: 0.0190 (0.0374)  time: 0.6312  data: 0.0140  max mem: 5493
Epoch: [0]  [200/400]  eta: 0:02:05  lr: 0.001000  loss: 0.7143 (0.9646)  loss_classifier: 0.3131 (0.5195)  loss_box_reg: 0.3700 (0.3663)  loss_objectness: 0.0296 (0.0466)  loss_rpn_box_reg: 0.0227 (0.0321)  time: 0.6147  data: 0.0113  max mem: 5495
Epoch: [0]  [300/400]  eta: 0:01:02  lr: 0.001000  loss: 0.6995 (0.8712)  loss_classifier: 0.2890 (0.4415)  loss_box_reg: 0.3290 (0.3512)  loss_objectness: 0.0183 (0.0454)  loss_rpn_box_reg: 0.0234 (0.0331)  time: 0.6101  data: 0.0114  max mem: 5505
Epoch: [0]  [399/400]  eta: 0:00:00  lr: 0.001000  loss: 0.6547 (0.8177)  loss_classifier: 0.2791 (0.4021)  loss_box_reg: 0.2963 (0.3406)  loss_objectness: 0.0218 (0.0425)  loss_rpn_box_reg: 0.0219 (0.0325)  time: 0.6143  data: 0.0119  max mem: 5505

IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.157
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.333
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.123
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.096
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.165
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.195
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.146
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.331
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.357
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.309
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.350
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.409

So I am getting slightly different results. Is anyone facing the same problem? And if so is this a Cuda related problem or does something in the model avoid the deterministic behavior?

Thanks in advance!
xXBasti

**Comment by fmassa** [comment]
Hi,

The model is not deterministic. Even if all the seeds are fixed, `RoIAlign` uses `atomicAdd` on the GPU, which breaks determinism, see https://github.com/pytorch/vision/blob/15e24bdfe0ba04d728afcff005ff2e01f1da9b33/torchvision/csrc/cuda/ROIAlign_cuda.cu#L286-L293

But the general question of wether we should have deterministic guarantees is a valid one, and there is some discussion in https://github.com/pytorch/pytorch/issues/15359 .

This is probably something easy to enforce in general, so I'm closing this issue for now. Let me know if you have other questions

**Comment by xXBasti** [comment]
Hi fmassa,
thanks for your reply and the useful information. 

Its seems that the discussion, points to an opinion against the feature :/

> This is probably something easy to enforce in general, so I'm closing this issue for now. Let me know if you have other questions

Would a solution be to force the calculation of RoIAlign to the CPU an use the CPU version of RoIAlign. Or is it possible the change the GPU execution to the CPU version or does the cause race conditions without a scheduling etc.?:
[https://github.com/pytorch/vision/blob/15e24bdfe0ba04d728afcff005ff2e01f1da9b33/torchvision/csrc/cpu/ROIAlign_cpu.cpp#L366-L371](https://github.com/pytorch/vision/blob/15e24bdfe0ba04d728afcff005ff2e01f1da9b33/torchvision/csrc/cpu/ROIAlign_cpu.cpp#L366-L371)

Thanks in advance!

**Comment by fmassa** [comment]
@xXBasti the problem lies not only in MaskRCNN per se, but many functions in PyTorch are also not deterministic.

> This is probably something easy to enforce in general, so I'm closing this issue for now. Let me know if you have other questions

I'm sorry, I just realized that I missed a **not** in the sentence, I meant

> This is probably **not** something easy to enforce in general

and I've updated my comment above.

For RoIAlign in particular, if you move the computation to the CPU that part would be deterministic I believe, but the slowdown would probably be severe.

Making `RoIAlign` backward deterministic would require rewriting a big part of it I'm afraid.

---

Also, there might be other parts of the model which uses operations that are not deterministic (specially in the backward), so this is a big endeavor, and I'm not sure it's currently worth it, given that many parts of PyTorch are not deterministic either.

**Comment by Cknex** [comment]
@fmassa  Hi, would it be possible to perform backward operation multiple times, and calculate the average? I think the average gradient might be more stable.

**Comment by yshvrdhn** [comment]
any updates on faster rcnn being deterministic?

**Comment by yshvrdhn** [comment]
@fmassa any updates on determinism in pytorch training for fastercnn and maskrcnn ?

**Comment by yshvrdhn** [comment]
Any updates here since we have moved to PyTorch 2.0.0 or we should still manually do the change in roi align ?
